File: resources/scop_1.5.1/examples/multi_listener.cpp
// multi_listener.cpp - DMI - 7-9-02

/* Usage: multi_listener [ <source-one> <source-two> ]
   (default sources are "news" and "updates") */

#include <sys/select.h>

#include <scop.h>

int main(int argc, char **argv)
{
	int sock[2];
	char *msg;
	fd_set read_fds;
	int max_fd;
	
	for(int i = 0; i < 2; i++)
		sock[i] = scop_open("localhost", "multi_listener");
	scop_listen(sock[0], argc == 3 ? argv[1] : (char *)"news");
	scop_listen(sock[1], argc == 3 ? argv[2] : (char *)"updates");
	
	while(1)
	{
		FD_ZERO(&read_fds);
		max_fd = 0;
		for(int i = 0; i < 2; i++)
		{
			FD_SET(sock[i], &read_fds);
			if(sock[i] > max_fd) max_fd = sock[i];
		}
		select(max_fd + 1, &read_fds, NULL, NULL, NULL);
		
		for(int i = 0; i < 2; i++)
		{
			if(FD_ISSET(sock[i], &read_fds))
			{		
				msg = scop_get_message(sock[i]);
				printf("Received <%s> from %s\n", msg,
						i == 1 ? "updates" : "news");
				delete[] msg;
			}
		}
	}
	
	return 0;
}

File: resources/scop_1.5.1/python/examples/multi_listener.py
#!/usr/bin/python
# multi_listener.cpp - DMI - 7-11-03

# Usage: multi_listener [ <source-one> <source-two> ]
#    (default sources are "news" and "updates")

import scop, sys, select

sock = []
for i in range(2):
	sock.append(scop.scop_open("localhost", "multi_listener"))
	
if len(sys.argv) == 3:
	scop.scop_listen(sock[0], argv[1])
	scop.scop_listen(sock[1], argv[2])
else:
	scop.scop_listen(sock[0], "news")
	scop.scop_listen(sock[1], "updates")

while 1:
	read_fds = [sock[0], sock[1]]
	r, w, e = select.select(read_fds, [], [])
	for fd in r:
		msg, rpc_flag = scop.scop_get_message(fd)
		print "Received <" + msg + "> from ",
		if fd == sock[1]:
			print "updates"
		else:
			print "news"

